File: ios/CaboGame/Models/GameState.swift
import Foundation

struct KingSwapHighlight: Codable, Equatable {
    var fromPlayerID: UUID
    var fromHandIndex: Int
    var toPlayerID: UUID
    var toHandIndex: Int
}

struct GameState: Codable {
    var players: [Player]
    var deck: [Card]
    var discardPile: [Card]
    var currentPlayerIndex: Int
    var phase: TurnPhase
    var pendingDraw: PendingDraw?
    var winnerID: UUID?
    var caboCallerID: UUID?
    var rematchRequestedByHost: Bool
    var readyPlayerIDs: Set<UUID>
    var hasStarted: Bool
    // Initial peek tracking per player (each player can peek up to 2 cards during initialPeek).
    var initialPeekedIndicesByPlayerIndex: [[Int]]
    var playersFinishedInitialPeek: Int
    var initialPeekGraceEndsAt: Date?
    /// Wall-clock deadline after which the current player's turn auto-expires.
    /// `nil` outside of main play (initial peek phase, game over, or before
    /// the first turn begins).
    var currentTurnEndsAt: Date?
    /// When non-nil, clients outline these two hand slots after a King swap.
    var kingSwapHighlight: KingSwapHighlight?

    init(players: [Player] = []) {
        self.players = players
        self.deck = .caboDeck()
        self.discardPile = []
        self.currentPlayerIndex = 0
        self.phase = .waitingForDraw
        self.pendingDraw = nil
        self.winnerID = nil
        self.caboCallerID = nil
        self.rematchRequestedByHost = false
        self.readyPlayerIDs = []
        self.hasStarted = false
        self.initialPeekedIndicesByPlayerIndex = Array(repeating: [], count: players.count)
        self.playersFinishedInitialPeek = 0
        self.initialPeekGraceEndsAt = nil
        self.currentTurnEndsAt = nil
        self.kingSwapHighlight = nil
    }

    var currentPlayerID: UUID? {
        guard players.indices.contains(currentPlayerIndex) else { return nil }
        return players[currentPlayerIndex].id
    }
}

File: android/app/src/main/kotlin/com/navitech/cabo/models/GameState.kt
package com.navitech.cabo.models

import kotlinx.serialization.Serializable

@Serializable
data class PendingDraw(val card: Card, val source: DrawSource)

@Serializable
data class KingSwapHighlight(
    val fromPlayerID: String,
    val fromHandIndex: Int,
    val toPlayerID: String,
    val toHandIndex: Int
)

@Serializable
data class GameState(
    val players: List<Player> = emptyList(),
    val deck: List<Card> = emptyList(),
    val discardPile: List<Card> = emptyList(),
    val currentPlayerIndex: Int = 0,
    val phase: TurnPhase = TurnPhase.WAITING_FOR_DRAW,
    val pendingDraw: PendingDraw? = null,
    val winnerID: String? = null,
    val caboCallerID: String? = null,
    val rematchRequestedByHost: Boolean = false,
    val readyPlayerIDs: Set<String> = emptySet(),
    val hasStarted: Boolean = false,
    // Initial peek tracking per player (up to 2 cards each during INITIAL_PEEK).
    val initialPeekedIndicesByPlayerIndex: List<List<Int>> = emptyList(),
    val playersFinishedInitialPeek: Int = 0,
    val initialPeekGraceEndsAt: Long? = null,  // epoch millis
    /** Epoch-millis deadline after which the current player's turn auto-expires.
     *  Null outside of main play (initial peek phase or game over). */
    val currentTurnEndsAt: Long? = null,
    /** When non-null, the UI outlines these two hand slots after a King swap (synced for all clients). */
    val kingSwapHighlight: KingSwapHighlight? = null
) {
    val currentPlayerID: String? get() = players.getOrNull(currentPlayerIndex)?.id
}
